src/oidcendpoint/session/grant.py

```diff
@@ -0,0 +1,337 @@
+        if iss.id == id:
+            return iss
+    return None
+
+
+TOKEN_MAP = {
+    "authorization_code": AuthorizationCode,
+    "access_token": AccessToken,
+    "refresh_token": RefreshToken
+}
+
+
+class Grant(Item):
+    attributes = ["scope", "claim", "resources", "authorization_details",
+                  "issued_token", "usage_rules", "revoked", "issued_at",
+                  "expires_at", "sub", "authorization_request",
+                  "authentication_event"]
+    type = "grant"
+
+    def __init__(self,
+                 scope: Optional[list] = None,
+                 claims: Optional[dict] = None,
+                 resources: Optional[list] = None,
+                 authorization_details: Optional[dict] = None,
+                 authorization_request: Optional[Message] = None,
+                 authentication_event: Optional[AuthnEvent] = None,
+                 issued_token: Optional[list] = None,
+                 usage_rules: Optional[dict] = None,
+                 issued_at: int = 0,
+                 expires_in: int = 0,
+                 expires_at: int = 0,
+                 revoked: bool = False,
+                 token_map: Optional[dict] = None,
+                 sub: Optional[str] = ""):
+        Item.__init__(self, usage_rules=usage_rules, issued_at=issued_at,
+                      expires_in=expires_in, expires_at=expires_at, revoked=revoked)
+        self.scope = scope or []
+        self.authorization_details = authorization_details or None
+        self.authorization_request = authorization_request or None
+        self.authentication_event = authentication_event or None
+        self.claims = claims or {}  # default is to not release any user information
+        self.resources = resources or []
+        self.issued_token = issued_token or []
+        self.id = uuid1().hex
+        self.sub = sub
+
+        if token_map is None:
+            self.token_map = TOKEN_MAP
+        else:
+            self.token_map = token_map
+
+    def get(self) -> object:
+        return GrantMessage(scope=self.scope, claims=self.claims,
+                            authorization_details=self.authorization_details,
+                            resources=self.resources)
+
+    def to_json(self) -> str:
+        d = {
+            "type": self.type,
+            "scope": self.scope,
+            "sub": self.sub,
+            "authorization_details": self.authorization_details,
+            "claims": self.claims,
+            "resources": self.resources,
+            "issued_at": self.issued_at,
+            "not_before": self.not_before,
+            "expires_at": self.expires_at,
+            "revoked": self.revoked,
+            "issued_token": [t.to_json() for t in self.issued_token],
+            "id": self.id,
+            "token_map": {k: ".".join([v.__module__, v.__name__]) for k, v in
+                          self.token_map.items()}
+        }
+        if isinstance(self.authorization_request, Message):
```

Comment:
Could we add the `authorization_request` and `authentication_event` to the dict above? So that it works when these objects are dicts too.

Reply:
I've taken care of the 2 cases (Message type or dictionary).